### PR TITLE
[AIR] `SklearnTrainer`&`Predictor` implementation

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Implements `SklearnTrainer` and `SklearnPredictor`. Full parallelism with joblib + support for GPU enabled estimators like cuML.

Interface has been modified slightly by addition of several arguments, which were required for full functionality.

I haven't tested cuML yet, will do it later.

Depends on https://github.com/ray-project/ray/pull/23889

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by amogkam** [comment]
@Yard1 would it be possible to split this up into smaller PRs (maybe one for ScalingConfig validation, one for the trainer, and one for the predictor)? Thanks!

**Comment by Yard1** [comment]
@amogkam the vast majority of this PR is the Trainer, with other changes being quite straightforward - I don't think splitting is going to have that much utility. Happy to do it if you (or someone else) insists, though!

**Comment by amogkam** [comment]
I think we should split out the config validation at least, since we need to do that for the other Trainers anyways.

File: python/ray/ml/config.py [hunk 1]
```diff
@@ -105,6 +115,50 @@ def as_placement_group_factory(self) -> "PlacementGroupFactory":
...
+        differs from the default value.
+
+        Args:
+            config: Config to check.
+            allowed_keys: ScalingConfigDataClass attribute keys that
+                can have a value different than the default one.
+            scaling_config_arg_name: Name of the ScalingConfig argument to be used
+                in the exception message.
+            exc_obj_name: Name of the object calling this method to be used
+                in the exception message.
```
Comment (krfricke) [inline python/ray/ml/config.py:137]:
```suggestion
            caller_name: Name of the object calling this method to be used
                in the exception message.
```

File: python/ray/ml/config.py [hunk 2]
```diff
@@ -105,6 +115,50 @@ def as_placement_group_factory(self) -> "PlacementGroupFactory":
         bundles = trainer_bundle + worker_bundles
         return PlacementGroupFactory(bundles, strategy=self.placement_strategy)
 
+    @classmethod
+    def validate_config(
+        cls,
+        config: Union[ScalingConfig, "ScalingConfigDataClass"],
+        allowed_keys: Iterable[str],
+        scaling_config_arg_name: str,
+        caller_name: str,
```
Comment (krfricke) [inline python/ray/ml/config.py:124]: Can we get rid of this? We shouldn't add arguments just for a print output. Instead, the enclosing function should raise the respective error. That function can just state the allowed keys. Using `raise from` will preserve information about which key is affected in the stacktrace.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 3]
```diff
@@ -63,7 +86,10 @@ class SklearnTrainer(Trainer):
             one is provided. All non-training datasets will be used as separate
             validation sets, each reporting separate metrics.
         label_column: Name of the label column. A column with this name
-            must be present in the training dataset.
+            must be present in the training dataset. If None, no validation
+            will be done.
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:90]:
```suggestion
            will be performed.
```

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 4]
```diff
@@ -112,26 +152,272 @@ def __init__(
...
         label_column: Optional[str] = None,
+        params: Optional[Dict[str, Any]] = None,
         scoring: Optional[ScoringType] = None,
         cv: Optional[CVType] = None,
+        parallelise_cv: Optional[bool] = None,
+        return_train_score_cv: bool = False,
         scaling_config: Optional[ScalingConfig] = None,
         run_config: Optional[RunConfig] = None,
         preprocessor: Optional[Preprocessor] = None,
-        resume_from_checkpoint: Optional[Checkpoint] = None,
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:120]: Can we still check if this is set below? I.e. `if fit_params.pop("resume_from_checkpoint", None): raise ValueError("SklearnTrainer does not support resuming from checkpoints...")`

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 5]
```diff
@@ -112,26 +152,272 @@ def __init__(
         estimator: BaseEstimator,
         datasets: Dict[str, GenDataset],
         label_column: Optional[str] = None,
+        params: Optional[Dict[str, Any]] = None,
         scoring: Optional[ScoringType] = None,
         cv: Optional[CVType] = None,
+        parallelise_cv: Optional[bool] = None,
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:158]: I'm a big fan of British English but we should probably keep this consistent across the library
```suggestion
        parallelize_cv: Optional[bool] = None,
```

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 6]
```diff
@@ -112,26 +152,272 @@ def __init__(
...
+                "for the dataset."
+            )
+        if (
+            not isinstance(self.parallelise_cv, bool)
+            and self.parallelise_cv is not None
+        ):
+            raise ValueError(
+                "`parallelise_cv` must be a bool or None, got "
+                f"'{self.parallelise_cv}'"
+            )
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:200]: What about validating `estimator`, `label_column`, `params`, `scoring`, `cv` and `return_train_score_cv`?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 7]
```diff
@@ -112,26 +152,272 @@ def __init__(
...
+                "trainer."
+            )
+
+    def _set_estimator_cpus(self, estimator: BaseEstimator, num_cpus: int) -> bool:
+        """Returns True if there are any params related to CPUs."""
+        cpu_params = {
+            param: num_cpus
+            for param in estimator.get_params(deep=True)
+            if param.endswith("n_jobs") or param.endswith("thread_count")
+        }
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:224]: How many of these parameter do we usually have? Aren't we overassigning CPUs here if we have more than one?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 8]
```diff
@@ -112,26 +152,272 @@ def __init__(
...
+                    else {"CPU": 1}
+                )
+                return PlacementGroupFactory([head_resources])
+
+            def _postprocess_checkpoint(self, checkpoint_path: str):
+                preprocessor = self._merged_config.get("preprocessor", None)
+                if not checkpoint_path or preprocessor is None:
+                    return
+                with open(os.path.join(checkpoint_path, PREPROCESSOR_KEY), "wb") as f:
+                    cpickle.dump(preprocessor, f)
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:421]: Should this be a method of the base Trainer class btw?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 9]
```diff
@@ -112,26 +152,272 @@ def __init__(
...
+        )
+        cv_aggregates = {}
+        for k, v in cv_results.items():
+            if not isinstance(v, np.ndarray):
+                continue
+            try:
+                cv_aggregates[f"{k}_mean"] = np.mean(v)
+                cv_aggregates[f"{k}_std"] = np.std(v)
+            except Exception:
+                # if we can't average for some reason
```
Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:301]: Should we warn here?

Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:301]: Should we warn here?

Comment (krfricke) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:301]: Should we warn here?

**Review by krfricke — CHANGES_REQUESTED** [review]
Looks great so far! Left a feq comments. Also let's make sure to replace all occurrences of "parallelise" with "parallelize"

**Comment by Yard1** [comment]
@amogkam @krfricke `validate_config` has been split into a separate PR - https://github.com/ray-project/ray/pull/23889

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 7 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:224]: Usually we will only have one `n_jobs`. However, in a case a user passes a sklearn Pipeline, each element of it may have its own `n_jobs` (`element__n_jobs`). Pipelines are ran sequentially, so there isn't a real possibility of overassignment.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 6 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:200]: We don't really validate similar arguments in other trainers - there will be an exception raised by sklearn during training if either of those is incorrect.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 8 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:431]: If we decide on it, then it will be a separate PR anyway, so let's not block on this

File: python/ray/ml/examples/sklearn_example.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    print(result.metrics)
+
+    return result
+
+
+def predict_sklearn(result: Result, num_workers: int, use_gpu: bool = False):
+    _, _, test_dataset = prepare_data()
+    checkpoint_object_ref = result.checkpoint.to_object_ref()
+
+    class SklearnScorer:
```
Comment (Yard1) [inline python/ray/ml/examples/sklearn_example.py:79]: Should use batch predictor

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 11]
```diff
@@ -112,26 +152,282 @@ def __init__(
...
+                    # parallelize CV
+                    n_jobs=1 if not parallelize_cv else num_cpus,
+                )
+            else:
+                validation_set_scores = {}
+                cv_scores = {}
+
+        results = {
+            **validation_set_scores,
+            **cv_scores,
```
Comment (gjoliver) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:396]: keys from validation_set_scores and cv_scores will never override each other?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 12]
```diff
@@ -112,26 +152,282 @@ def __init__(
...
+                "`parallelize_cv` cannot be True if there are GPUs assigned to the "
+                "trainer."
+            )
+
+    def _set_estimator_cpus(self, estimator: BaseEstimator, num_cpus: int) -> bool:
+        """Returns True if there are any params related to CPUs."""
+        cpu_params = {
+            param: num_cpus
+            for param in estimator.get_params(deep=True)
+            if param.endswith("n_jobs") or param.endswith("thread_count")
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:229]: This is the intended way.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 11 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:396]: Never

File: python/ray/ml/examples/sklearn_example.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+    # Scale some random columns
+    columns_to_scale = ["mean radius", "mean texture"]
+    preprocessor = Chain(
+        OrdinalEncoder(["categorical_column"]), StandardScaler(columns=columns_to_scale)
+    )
+
+    if use_gpu:
+        trainer_resources = {"CPU": 1, "GPU": 1}
+        estimator = cuMLRandomForestClassifier()
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/examples/sklearn_example.py:54]: Yes, it will only use 1 CPU and 1 GPU.

File: python/ray/ml/examples/sklearn_example.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+try:
+    from cuml.ensemble import RandomForestClassifier as cuMLRandomForestClassifier
+except ImportError:
+    cuMLRandomForestClassifier = None
+
+
+def prepare_data() -> Tuple[Dataset, Dataset, Dataset]:
+    data_raw = load_breast_cancer()
+    dataset_df = pd.DataFrame(data_raw["data"], columns=data_raw["feature_names"])
+    dataset_df["target"] = data_raw["target"]
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/examples/sklearn_example.py:27]: Sorry, can you clarify? What are you referring to?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 15]
```diff
@@ -112,26 +152,282 @@ def __init__(
...
+                    else {"CPU": 1}
+                )
+                return PlacementGroupFactory([head_resources])
+
+            def _postprocess_checkpoint(self, checkpoint_path: str):
+                preprocessor = self._merged_config.get("preprocessor", None)
+                if not checkpoint_path or preprocessor is None:
+                    return
+                with open(os.path.join(checkpoint_path, PREPROCESSOR_KEY), "wb") as f:
+                    cpickle.dump(preprocessor, f)
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:431]: this checkpoint is mostly for fault tolerance purpose right?
we are not trying to persist and use it across runs for example?

[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:431]: We do not support resuming training from checkpoints, but we need it for prediction. This is the same logic as in eg. `XGBoostTrainer`.

File: python/ray/ml/examples/sklearn_example.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply python/ray/ml/examples/sklearn_example.py:27]: oh I guess you are trying to merge data_raw["target"] into data_raw["data"] as a feature column.
hmm, ok, nevermind.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 7 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py]: Made the behavior toggelable

File: python/ray/ml/config.py [hunk 16]
```diff
@@ -1,5 +1,15 @@
 from dataclasses import dataclass
-from typing import Dict, Any, Optional, List, Mapping, Callable, Union, TYPE_CHECKING
+from typing import (
+    Dict,
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/config.py:3]: will be done in https://github.com/ray-project/ray/pull/23889

File: python/ray/ml/config.py [hunk 17]
```diff
@@ -105,6 +115,50 @@ def as_placement_group_factory(self) -> "PlacementGroupFactory":
...
+                )
+
+        prohibited_keys = set(default_config.__dict__) - allowed_keys
+
+        for key in prohibited_keys:
+            if config.__dict__[key] != default_config.__dict__[key]:
+                raise ValueError(
+                    f"Key '{key}' cannot be used with {caller_name}. Please remove it "
+                    f"from {scaling_config_arg_name} or set it to its default value of "
+                    f"'{default_config.__dict__[key]}'"
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/config.py:159]: will be done in https://github.com/ray-project/ray/pull/23889

File: python/ray/ml/examples/sklearn_example.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+import ray
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.predictors.integrations.sklearn import SklearnPredictor
+from ray.ml.train.integrations.sklearn import SklearnTrainer
+from ray.data.dataset import Dataset
+from ray.ml.result import Result
+from ray.ml.preprocessors import StandardScaler, OrdinalEncoder, Chain
+from sklearn.datasets import load_breast_cancer
+from sklearn.model_selection import train_test_split
+from sklearn.ensemble import RandomForestClassifier
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/examples/sklearn_example.py:16]: done

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 19]
```diff
@@ -112,26 +152,282 @@ def __init__(
...
+                logger.warning(
+                    f"Couldn't calculate aggregate metrics for CV folds! {e}"
+                )
+                cv_aggregates[f"{k}_mean"] = np.nan
+                cv_aggregates[f"{k}_std"] = np.nan
+                pass
+
+        return {"cv": {**cv_results, **cv_aggregates}}
+
+    def _set_estimator_cv_parallelism(self, num_cpus: int, has_gpus: bool) -> bool:
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py]: refactored this, PTAL

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 11 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:393]: done

File: python/ray/ml/config.py (hunk 2 above)
Comment (Yard1) [reply python/ray/ml/config.py:124]: will be done in https://github.com/ray-project/ray/pull/23889

File: python/ray/ml/config.py (hunk 1 above)
Comment (Yard1) [reply python/ray/ml/config.py:137]: will be done in https://github.com/ray-project/ray/pull/23889

File: python/ray/ml/config.py [hunk 20]
```diff
@@ -105,6 +115,50 @@ def as_placement_group_factory(self) -> "PlacementGroupFactory":
         bundles = trainer_bundle + worker_bundles
         return PlacementGroupFactory(bundles, strategy=self.placement_strategy)
 
+    @classmethod
+    def validate_config(
```
Comment (Yard1) [inline python/ray/ml/config.py:119]: Split into a separate PR - https://github.com/ray-project/ray/pull/23889

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 21]
```diff
@@ -112,26 +157,273 @@ def __init__(
...
+        datasets = self._get_datasets()
+        X_train, y_train = datasets.pop(TRAIN_DATASET_KEY)
+        groups = None
+        if "cv_groups" in X_train.columns:
+            groups = X_train["cv_groups"]
+            X_train = X_train.drop("cv_groups", axis=1)
+
+        trial_resources = tune.get_trial_resources().required_resources
+        num_cpus = int(trial_resources.get("CPU", 0))
+        has_gpus = bool(trial_resources.get("GPU", 0))
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:349]: Can we get the resources via `self.scaling_config` instead? `tune.get_trial_resources` is not a public API, and I think we should stick to using the scaling config abstraction instead of PlacementGroupFactory.

`self.scaling_config` is already updated based on any provided hyperparameters.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 22]
```diff
@@ -112,26 +157,273 @@ def __init__(
...
+            set_cpu_params(self.estimator, num_estimator_cpus)
+
+        with parallel_backend("ray", n_jobs=num_cpus):
+            start_time = time()
+            self.estimator.fit(X_train, y_train, **self.fit_params)
+            fit_time = time() - start_time
+
+            with tune.checkpoint_dir(step=1) as checkpoint_dir:
+                with open(os.path.join(checkpoint_dir, MODEL_KEY), "wb") as f:
+                    cpickle.dump(self.estimator, f)
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:369]: Can we add the preprocessor directly here instead of in `_postprocess_checkpoint`? We should work towards removing `_postprocess_checkpoint` I feel.

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 23]
```diff
@@ -112,26 +157,273 @@ def __init__(
...
+        scaling_config = self.scaling_config
+
+        class SklearnTrainable(trainable_cls):
+            @classmethod
+            def default_resource_request(cls, config):
+                updated_scaling_config = config.get("scaling_config", scaling_config)
+                updated_scaling_config = ScalingConfigDataClass(
+                    **updated_scaling_config
+                )
+                ScalingConfigDataClass.validate_config(
```
Comment (amogkam) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:408]: See the comments in the other PR, but I would say let's move this validation to the base Trainer class and have the allowed_keys as class attributes. 

If we do that, and then also remove `_postprocess_checkpoint`, then this subclass should not need to override `as_trainable`. And if it's still necessary, that means we need to rethink our original API.

**Review by amogkam — CHANGES_REQUESTED** [review]
LGTM overall! Just left some small comments.

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py [hunk 24]
```diff
@@ -36,12 +45,24 @@ def from_checkpoint(cls, checkpoint: Checkpoint) -> "SklearnPredictor":
...
+        estimator_path = os.path.join(path, MODEL_KEY)
+        with open(estimator_path, "rb") as f:
+            estimator = cpickle.load(f)
+        preprocessor_path = os.path.join(path, PREPROCESSOR_KEY)
+        if os.path.exists(preprocessor_path):
+            with open(preprocessor_path, "rb") as f:
+                preprocessor = cpickle.load(f)
+        else:
+            preprocessor = None
+        shutil.rmtree(path)
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:58]: I was going to suggest that as well! Would be quite useful. `to_directory` could also take in an argument to avoid making a temp copy if the checkpoint is already a directory.

File: python/ray/ml/examples/sklearn_example.py [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+    _, _, test_dataset = prepare_data()
+
+    batch_predictor = BatchPredictor.from_checkpoint(
+        result.checkpoint, SklearnPredictor
+    )
+
+    predicted_labels = (
+        batch_predictor.predict(
+            test_dataset,
+            num_gpus_per_worker=int(use_gpu),
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/examples/sklearn_example.py:87]: Yes

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py [hunk 26]
```diff
@@ -100,4 +124,24 @@ def predict(
...
+
+        if num_estimator_cpus:
+            set_cpu_params(self.estimator, num_estimator_cpus)
+
+        if feature_columns:
+            if isinstance(data, np.ndarray):
+                data = data[:, feature_columns]
+            else:
+                data = data[feature_columns]
+        with parallel_backend("ray", n_jobs=num_estimator_cpus):
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:140]: This will be set to `n_jobs=1` by default, which means it will fallback to joblib's SequentialBackend no matter what's set here. Unless the user specifically sets `num_estimator_cpus`, there will be no parallelism.

File: python/ray/ml/utils/sklearn_utils.py [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+        param: num_cpus
+        for param in estimator.get_params(deep=True)
+        if any(
+            param.endswith(cpu_param_name) for cpu_param_name in SKLEARN_CPU_PARAM_NAMES
+        )
+    }
+    estimator.set_params(**cpu_params)
+
+
+# score func from https://github.com/scikit-learn/scikit-learn/blob/\
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/utils/sklearn_utils.py:35]: Signature differs between releases

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 22 above)
Comment (Yard1) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:369]: Yes, we should be able to do it here

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:58]: `with checkpoint.to_temp_directory() as tempdir` sounds useful.

I wouldn't go with the argument here - while it is nice to avoid inspecting internal state (e.g. in HF trainer checking the `_local_path` attribute) we don't know if we should clean up the returned path or not as we don't know if it's a temporary path or the existing (potentially durable) path. 

Actually, `with checkpoint.to_temp_directory() as tempdir` could just return the durable path if it's already a directory checkpoint. Probably should have a different name then, though.

[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:58]: Suggested as `Checkpoint.as_directory()`: https://github.com/ray-project/ray/pull/23908/files

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py (hunk 6 above)
Comment (krfricke) [reply python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:213]: I think it's nice to do the validation before we kick off a distributed run, though, so we may want to do this for all trainers. I can update this when taking over

File: python/ray/ml/utils/sklearn_utils.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/ml/utils/sklearn_utils.py:35]: Do we need to keep this in sync and how is the process for that?

File: python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/predictors/integrations/sklearn/sklearn_predictor.py:140]: Got it. Can we add a test for full batch inference with num_estimator_cpus greater than 1?

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 28]
```diff
@@ -95,43 +120,303 @@ class SklearnTrainer(Trainer):
...
+        scaling_config_dataclass = self._validate_and_get_scaling_config_data_class(
+            self.scaling_config
+        )
+
+        num_workers = scaling_config_dataclass.num_workers or 0
+        assert num_workers == 0  # num_workers is not in scaling config allowed_keys
+        has_gpus = scaling_config_dataclass.use_gpu
+
+        trainer_resources = scaling_config_dataclass.trainer_resources or {"CPU": 1}
+        num_cpus = trainer_resources.get("CPU", 1.0)
```
Comment (Yard1) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:370]:
```suggestion
        num_cpus = int(trainer_resources.get("CPU", 1)) 
```
Has to be an int, otherwise sklearn will throw an exception in some estimators

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 29]
```diff
@@ -344,19 +358,16 @@ def training_loop(self) -> None:
...
             X_train = X_train.drop("cv_groups", axis=1)
 
-        scaling_config_dataclass = ScalingConfigDataClass(**self.scaling_config)
+        scaling_config_dataclass = self._validate_and_get_scaling_config_data_class(
+            self.scaling_config
+        )
 
         num_workers = scaling_config_dataclass.num_workers or 0
+        assert num_workers == 0  # num_workers is not in scaling config allowed_keys
         has_gpus = scaling_config_dataclass.use_gpu
```
Comment (Yard1) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:367]:
```suggestion
```
`use_gpu` is a prohibited key and it only refers to workers

File: python/ray/ml/train/integrations/sklearn/sklearn_trainer.py [hunk 30]
```diff
@@ -344,19 +358,16 @@ def training_loop(self) -> None:
...
 
         trainer_resources = scaling_config_dataclass.trainer_resources or {"CPU": 1}
-        worker_resources = scaling_config_dataclass.resources_per_worker or {
-            "CPU": 1,
-            "GPU": int(has_gpus),
-        }
-        num_cpus = trainer_resources.get(
-            "CPU", 1.0
-        ) + num_workers or 0 * worker_resources.get("CPU", 1.0)
+        num_cpus = int(trainer_resources.get("CPU", 1))
```
Comment (Yard1) [inline python/ray/ml/train/integrations/sklearn/sklearn_trainer.py:370]:
```suggestion
        has_gpus = bool(trainer_resources.get("GPU", 0))
        num_cpus = int(trainer_resources.get("CPU", 1))
```

**Review by Yard1 — COMMENTED** [review]
Thanks for the help @krfricke! Spotted one oversight regarding GPU use

File: python/ray/ml/tests/test_sklearn_predictor.py [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+        checkpoint = Checkpoint.from_directory(tmpdir)
+        predictor = SklearnPredictor.from_checkpoint(checkpoint)
+
+    data_batch = np.array([[1, 2], [3, 4], [5, 6]])
+    predictions = predictor.predict(data_batch)
+
+    assert len(predictions) == 3
+
+
+def test_batch_prediction_with_set_cpus(ray_start_4_cpus):
```
Comment (amogkam) [inline python/ray/ml/tests/test_sklearn_predictor.py:125]: Love this test!

File: python/ray/ml/utils/sklearn_utils.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/ml/utils/sklearn_utils.py:35]: Should we just have conditional logic to check the sklearn version and call the appropriate API as necessary? I have the same concern as @krfricke for future maintainability.

**Review by amogkam — COMMENTED** [review]
LGTM overall! Just left one comment on copying over the `score` function- is there no way to avoid that?

File: python/ray/ml/utils/sklearn_utils.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/utils/sklearn_utils.py]: @amogkam removed it

**Review by amogkam — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
